### ISSUE TITLE
Design new ORM storage

### ISSUE THREAD

**Issue by samselikoff** [samselikoff/ember-cli-mirage#886]
Need to think through how the new underlying storage for the ORM needs to work to support one-sided relationships.

**Comment by samselikoff** [comment]
# The problem

The key difference is that the new storage engine must support one-way relationships.

The current store, which supports one-to-many relationships, sets up the database using traditional foreign keys. So, if you define a post that belongs to a user:

```js
// models/post.js
Model.extend({
  
  user: belongsTo()
  
})
```

what this will do is add a `userId` foreign key to the `posts` database collection, since this is traditionally how one-to-many relationships are modeled.

If we define a post that has many tags:

```js
// models/post.js
Model.extend({
  
  tags: hasMany()
  
})
```

we do a similar thing: we add a foreign key to the `tag` model (the other side of the hasMany). But typically, blog posts and tags have a many-to-many relationship - so this assumption is wrong. However, what about comments?

```js
// models/post.js
Model.extend({
  
  comments: hasMany()
  
})
```

In this case, each comment most likely does have a `postId`, so the assumption of one-to-many is correct. So, the initial thought was to make assumptions, and provide a way for users to override those assumptions if they turned out to be wrong.

However, APIs will often expose something like the above post model that has many tags, without tags ever pointing back to posts:

```
GET /posts/1
{
  post: {
    id: 1,
    title: 'Lorem ipsum',
    tags_ids: [1, 2, 4]
  },
  tags: [
    { id: 1, name: 'Economics' },
    { id: 2, name: 'Philosophy' },
    { id: 4, name: 'Politics' }
  ]
}
```

Mirage as a tool has a single purpose: to improve the UX of the developer. If the API a developer happens to be working with has no concept of a relationship from one resource to another (in this case, from a tag back to a post), why should they need to waste brain cycles deciding upon and creating one in Mirage? Answer: they shouldn't.

As a result, the notion of single-owner foreign keys must be dropped from Mirage.

# Approach

As in Ember Data, when we set up the following post model

```js
// models/post.js
Model.extend({
  
  tags: hasMany()
  
})
```

all we want is a `tagIds` property to live somewhere. The main question is: where?

Currently, all data is "persisted" in the database. The database is a collection of `DbCollection` objects (that was creative). Currently, `hasMany` and `belongsTo` define special methods on models to manage relationships.

So, if you had a `post` with a `userId`, doing something like

```js
post.userId = 1;
```

would execute a magic method to make sure references were updated. Persisting this change would mean saving the `userId` of `1` back to the database.

The `hasMany` case is interesting because it also worked through the belongs-to side of the relationship, and its foreign key. Let's talk through what the has-may side should look like now, with one-sided relationships.

Using the post model above, I can think of a few places to store the `tagIds` property:

1. In the database, alongside each post "row"
2. In a new object, something like an `AssociationsMap`, which would serve to be a single global identity map for all association state.

## In the database

If we store it in the db, it looks like this:

```
posts: [
  { id: 1, title: 'Lorem ipsum', tagIds: [1, 4, 5] }
];
```

The magic method `post.tags` would use this data to find and instantiate the associated tag models.

This seems ok. What happens when we define the inverse?

```js
// models/tag.js
Model.extend({
  
  post: hasMany()
  
})
```

```
tags: [
  { id: 1, title: 'Economics', postIds: [1, 6, 7, 10, 11] }
]
```

We now have data in two places...which is a bummer. The methods `tag.postIds = [ ... ]` and `post.tagIds = [ ... ]` will have to sync with each other. Seems like we could get into trouble.

## In an identity map

What if instead we keep plain attributes in the database, but create a new object to store associations?

```js
// models/post.js
Model.extend({

  tags: hasMany()
  
});


// some-registry
_associations: {
  posts: [
    { id: 1, tagIds: [1, 4, 5] },
    { id: 2, tagIds: [1, 4, 5] }
  ]
}
```

Then when we come around and define the inverse

```js
// models/tag.js
Model.extend({
  
  post: hasMany()
  
})
```

we would start by considering that this `tag => post` relationship could have an inverse; check the `_associations` registry for a `post => tag` relationship; and if it exists, use and manage it instead.

The problem here is, how to fetch `tag.posts`. Looking at the data structure above, this would mean finding all `posts` where `tagIds.contains(id)`. Could do worse, I suppose, but it feels a bit smelly.

Alternatively we could store both sides of the relationship separately, which would simplify and speed up lookup, but we're back to having to keep the two in sync.

```js
// some-registry
_associations: {
  posts: [
    { id: 1, tagIds: [1, 4, 5] },
    { id: 2, tagIds: [1, 4, 5] }
  ],
  tags: [
    { id: 1, postIds: [1, 2, 4] },
    { id: 1, postIds: [3, 5, 8] },
  ]
}
```

What if the data structure of the identity map were just an array of associations?

```
_associations: [
  { from: 'post', to: 'tag', postId: 1, tagId: 1 },
  { from: 'post', to: 'tag', postId: 1, tagId: 2 },
  { from: 'post', to: 'tag', postId: 1, tagId: 4 },
  { from: 'post', to: 'tag', postId: 1, tagId: 7 },
]
```

Basically, an array of join records. Additional properties could be used to support named, inverse, and polymorphic associations:

```
_associations: [
  { from: 'post', to: 'tag', name: 'blueTags', postId: 1, tagId: 1 },
  { from: 'post', to: 'tag', name: 'redTags', inverse: 'myPosts', postId: 1, tagId: 2 },
  { from: 'image', to: 'imageable', imageId: 1, imageableId: 4, type: 'post' }
]
```

but then every association query would need to traverse this array.

Other ideas/brainstorming:

```js
_associations: {
  refs: {
    posts: [
      { id: 1, tags: postsTags }
    ],
    tags: [
      { id: 1, posts: postsTags }
    ],
  },
  data: [
    postsTags: [
      { postId: 1, tagId: 1 },
      { postId: 1, tagId: 2 },
      { postId: 3, tagId: 1 },
      { postId: 2, tagId: 4 },
      { postId: 2, tagId: 5 }
    ]
  ]
}
```

```js
_associations: {
  post: {
    'blueComments': [1, 4, 5],
    'redComments': [2, 3]
  },
  comments: {
    
  }
}
```

# Other considerations

## Using a real database

If we go this route, and ever want to use an actual database to persist data (for purposes of development and testing), what are the implications? I believe an adapter would be sufficient to take the database + associations map and translate it to a data structure that's capable of being stored in a traditional database.

## Using fixtures / the db to set up state

Regardless of how the internal storage works, it's still nice to be able to use fixtures in the traditional database format to set up data:

```js
fixtures: {
  posts: [
    { id: 1, title: 'Lorem ipsum'] }
  ],
  comments: [
    { id: 1, text: 'Foo', postId: 1 }
  ]
}
```

and a lot of folks are doing this now. I think we can still have a layer that takes this format and puts it into the new storage format on init, but regardless it's something that needs to be considered.

**Comment by samselikoff** [comment]
Would love any feedback on this btw.

**Comment by ef4** [comment]
The storage format can be entirely internal implementation detail. There's no reason you can't keep using the foreign key implementation you already have, even for "one way" relationships. 

All that's need is conventions for deciding what parts of the internal storage to surface in the public API. One way to do that is to require users to declare both halves of a relationship If they want the accessors to work from both sides. Otherwise they get one sided accessors. But in either case the storage structure remains the same.

**Comment by samselikoff** [comment]
> One way to do that is to require users to declare both halves of a relationship If they want the accessors to work from both sides. Otherwise they get one sided accessors.

I want this, I think it's the most reasonable, it's how Active Record/Ember Data work, etc.

My problem is, if you start with one side:

```js
// post
Model.extend({

  comments: hasMany()

});
```

where do I store the `commentIds`? If I store them with the "record" in the existing `post` `DbCollection` as a plain array, how do I sync them up if/when the inverse relationships is defined on the comment?

**Comment by samselikoff** [comment]
Right now the implementation is to stick a fk on post, but that is a flawed assumption. Are you saying I could still do that, but just not expose that fk as an attribute, or to the serializer layer etc.?

**Comment by ef4** [comment]
Yes, exactly. The FK can always go on post. It can just stay private unless the user declares a two sided relationship.

**Comment by samselikoff** [comment]
I've decided to go with one-way associations, and the storage format will mimic Ember Data:

- belongsTo will get a `modelId` property
- hasMany will get a `modelIds` property

These won't touch any other models, and this will require us to "sync" the two when inverse relationships change. Note that even if we kept fk storage to a single owner (i.e. if the implementation of hasMany was to store a fk on the child), models themselves still have the potential to "fall out of sync" since they are not bound/observable. 

The main reason to not store the fk on the child is because it's simply an incorrect assumption in many cases, and it will lead to a more confusing codebase (and potentially confusion for end users, if the implementation ever leaks through). If a post hasMany tags, a post will get a tagIds property - and that's it. Easy to understand, both from the end-user API side and the internal implementation. There's no reason to create a tag.postId property, when one never exists. Further, there's no way to guess this is a many-to-many relationship, from just the one-sided hasMany declaration alone. Therefore, we will use a post.tagIds property, and manage/sync it behind the scenes. Users may have to call `model.reload()` to assert against "fresh" db state, just as they do today. I don't believe this has not caused many issues.
